### PR TITLE
docs(apisix): wrote transformation/transcode-http-to-grpc

### PR DESCRIPTION
**Pull request by An-DJ**
Preview: https://api7-docs-git-transform-grpc-api7ai.vercel.app/apisix/how-to-guide/transformation/transcode-http-to-grpc

### CODE REVIEW

**Comment by Boburmirzo** [comment]
It looks good to me.

**Comment by kayx23** [comment]
Round 1 review done. Confirmed that the code provided in the doc works as intended. 
Please see Vercel comments @An-DJ

File: apisix_versioned_docs/version-3.2.0/how-to-guide/transform/grpc.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
+---
+title: gRPC Proxy
+slug: /how-to-guide/transform/grpc
+---
+
+[gRPC](https://grpc.io/) is a open source high performance Remote Procedure Call (RPC) framework based on HTTP/2 protocol. It uses [protocol buffers](https://protobuf.dev/) (a.k.a. _protobuf_) as the interface description language. The gRPC service provides portable mechanism for load balancing, link tracing, health checking, and authentication, effectively connecting multiple services between datacenters.
+
+APISIX supports proxy gRPC using the plugin `grpc-transcode`, which invokes gRPC services in a RESTful way. The plugin supports specifying the contents of `.proto` files in Apache APISIX and implementing proxies for different gRPC services through user-defined gRPC services.
```
Comment (kayx23) [inline apisix_versioned_docs/version-3.2.0/how-to-guide/transform/grpc.md:8]:
```suggestion
APISIX supports transcoding between HTTP and gRPC using the plugin `grpc-transcode`. The plugin supports specifying the contents of `.proto` files in Apache APISIX and implementing proxies for different gRPC services through user-defined gRPC services.
```

File: apisix_versioned_sidebars/version-3.2.0-sidebars.json [hunk 2]
```diff
@@ -71,6 +71,13 @@
           "items": [
             "how-to-guide/authentication/sso-with-oidc-and-keycloak"
           ]
+        },
+        {
+          "type": "category",
+          "label": "Transform",
+          "items": [
+            "how-to-guide/transform/grpc"
```
Comment (kayx23) [inline apisix_versioned_sidebars/version-3.2.0-sidebars.json:79]:
```suggestion
          "label": "Transformation",
          "items": [
            "how-to-guide/transformation/grpc"
```

File: apisix_versioned_docs/version-3.2.0/how-to-guide/transform/grpc.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
+---
+title: Transcode HTTP to gRPC
+slug: /how-to-guide/transform/grpc
+---
+
+[gRPC](https://grpc.io/) is a open source high performance Remote Procedure Call (RPC) framework based on HTTP/2 protocol. It uses [protocol buffers](https://protobuf.dev/) (a.k.a. _protobuf_) as the interface description language. gRPC framework provides a portable mechanism for load balancing, link tracing, health checking, and authentication, effectively connecting multiple services between datacenters.
```
Comment (kayx23) [inline apisix_versioned_docs/version-3.2.0/how-to-guide/transform/grpc.md:6]:
```suggestion
[gRPC](https://grpc.io/) is an open-source high performance Remote Procedure Call (RPC) framework based on HTTP/2 protocol. It uses [protocol buffers](https://protobuf.dev/) (a.k.a. _protobuf_) as the interface description language. gRPC framework provides a portable mechanism for load balancing, link tracing, health checking, and authentication, effectively connecting multiple services between datacenters.
```

File: apisix_versioned_sidebars/version-3.2.0-sidebars.json (hunk 2 above)
Comment (An-DJ) [reply apisix_versioned_sidebars/version-3.2.0-sidebars.json]: Done in the latest commit.

File: apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    }
+  }
+}'
+```
+
+An `HTTP/1.1 502 Bad Gateway` response verifies that the plugin `grpc-transcode` is disabled successfully.
+
+## Next Steps
+
+APISIX also supports process gRPC Web requests from JavaScript clients to a gRPC service directly by another plugin `grpc-web`.
```
Comment (kayx23) [inline apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md:200]:
```suggestion

Learn more about the `grpc-transcode` plugin in the plugin reference (coming soon).

In addition to transcoding HTTP requests to gRPC requests, APISIX also supports [gRPC-Web](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md), a variation of the [native gRPC protocol](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md), with the APISIX plugin `grpc-web`. 

[//]: <TODO: Add the link to how-to guide or plugin's doc>
```

File: apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+[//]: <TODO: grpc-web plugin>
```
Comment (kayx23) [inline apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md:202]:
```suggestion
[//]: <TODO: grpc-transcode plugin reference doc>
[//]: <TODO: grpc-web plugin guide>
```

File: apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
+---
+title: Transcode HTTP to gRPC
+slug: /how-to-guide/transformation/grpc
+---
+
+[gRPC](https://grpc.io/) is an open-source high performance Remote Procedure Call (RPC) framework based on HTTP/2 protocol. It uses [protocol buffers (protobuf)](https://protobuf.dev/) as the interface description language (IDL). gRPC framework provides a portable mechanism for load balancing, link tracing, health checking, and authentication, effectively connecting multiple services between datacenters.
+
+APISIX supports transforming between HTTP RESTful and gRPC using the plugin `grpc-transcode`. The plugin supports specifying the contents of `.proto` files in Apache APISIX and implementing proxies for different gRPC services through user-defined gRPC services.
```
Comment (kayx23) [inline apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md:8]:
```suggestion
APISIX provides the capability to transform between RESTful HTTP and gRPC requests and responses, using the plugin `grpc-transcode` and proto objects. 
 
 [//]: <TODO: hyperlink proto>
```

File: apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+## Prerequisite(s)
+
+* Install [Docker](https://docs.docker.com/get-docker/).
+* Install [cURL](https://curl.se/) to send requests to APISIX for validation.
+* Install [gRPCurl](https://github.com/fullstorydev/grpcurl) to send requests to gRPC services for validation.
+* Follow the [Getting Started tutorial](./../../getting-started/) to start a new APISIX instance in Docker.
+
+## Deploy Example gRPC Server
+
+Spin up an [example gRPC server](https://github.com/api7/grpc_server_example) instance named `quickstart-grpc-example` in Docker, exposed on port `50051`:
```
Comment (kayx23) [inline apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md:27]:
```suggestion
Start an [example gRPC server](https://github.com/api7/grpc_server_example) Docker instance `quickstart-grpc-example` on port `50051`:
```

File: apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+<br/>
+
+## Prerequisite(s)
+
+* Install [Docker](https://docs.docker.com/get-docker/).
+* Install [cURL](https://curl.se/) to send requests to APISIX for validation.
+* Install [gRPCurl](https://github.com/fullstorydev/grpcurl) to send requests to gRPC services for validation.
+* Follow the [Getting Started tutorial](./../../getting-started/) to start a new APISIX instance in Docker.
+
+## Deploy Example gRPC Server
```
Comment (kayx23) [inline apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md:25]: how about
```suggestion
## Deploy a gRPC Server
```

File: apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+service EchoService {
+  rpc Echo (EchoMsg) returns (EchoMsg);
+}
+
+message EchoMsg {
+  string msg = 1;
+}
+```
+
+For example, the gRPC method `echo.EchoService.Echo` accept parameters of `EchoMsg` type, and respond with `EchoMsg` type also.
```
Comment (kayx23) [inline apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md:60]:
```suggestion
In this example, `Echo` is a method in `EchoService` that accepts parameter type `string` defined in `EchoMsg`.
```

File: apisix_versioned_sidebars/version-3.2.0-sidebars.json [hunk 10]
```diff
@@ -71,6 +71,13 @@
           "items": [
             "how-to-guide/authentication/set-up-sso-with-oidc-and-keycloak"
           ]
+        },
+        {
+          "type": "category",
+          "label": "Transformation",
+          "items": [
+            "how-to-guide/transformation/grpc"
```
Comment (kayx23) [inline apisix_versioned_sidebars/version-3.2.0-sidebars.json:79]:
```suggestion
            "how-to-guide/transformation/transcode-http-to-grpc"
```

File: apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    string msg = 1;
+  }"
+}'
+```
+
+An `HTTP/1.1 201 OK` response verifies that the proto object is created successfully.
+
+## Enable `grpc-transcode` Plugin
+
+Create a route with id `quickstart-grpc` and enable the plugin `grpc-transcode`. It forwards HTTP requests sent to `/echo` to the upstream gRPC server `quickstart-grpc-example` spun up before:
```
Comment (kayx23) [inline apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md:105]:
```suggestion
Create a route with id `quickstart-grpc` and enable the plugin `grpc-transcode`:
```

File: apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+}'
+```
+
+❶ `proto_id`: the proto object which defines gRPC services
+
+❷ `service`: gRPC service `echo.EchoService` in use
+
+❸ `method`: gRPC method `Echo` in use
+
+The route transforms HTTP Restful requests to gRPC requests and sends them to gRPC server's method `echo.EchoService/Echo`, then transforms the gRPC responses from gRPC server to HTTP Restful responses and sent them to clients.
```
Comment (kayx23) [inline apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md:141]:
```suggestion
APISIX now transcodes the RESTful HTTP requests received at `/echo` route to proto requests and forwards them to the upstream gRPC server `quickstart-grpc-example` to invoke the method `echo.EchoService/Echo`. Once the gRPC server responds, APISIX transcodes the proto responses back to RESTful HTTP responses for clients.
```

File: apisix_versioned_sidebars/version-3.2.0-sidebars.json (hunk 10 above)
Comment (An-DJ) [reply apisix_versioned_sidebars/version-3.2.0-sidebars.json]: Done.

**Review by kayx23 — APPROVED** [review]

**Pull request merged**